<h1><strong>Data preprocessing</strong></h1>

<h1>Team Science Bros</h1>
In this notebook file you shall find data preprocessing script.

In [ ]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np

In [ ]:
# Importing and reading the dataset as df 
df = pd.read_csv('/content/Stroke dataset.csv')
df

id  gender   age  hypertension  heart_disease ever_married  \
0      9046    Male  67.0             0              1          Yes   
1     51676  Female  61.0             0              0          Yes   
2     31112    Male  80.0             0              1          Yes   
3     60182  Female  49.0             0              0          Yes   
4      1665  Female  79.0             1              0          Yes   
...     ...     ...   ...           ...            ...          ...   
5105  18234  Female  80.0             1              0          Yes   
5106  44873  Female  81.0             0              0          Yes   
5107  19723  Female  35.0             0              0          Yes   
5108  37544    Male  51.0             0              0          Yes   
5109  44679  Female  44.0             0              0          Yes   

          work_type Residence_type  avg_glucose_level   bmi   smoking_status  \
0           Private          Urban             228.69  36.6  formerly smoked   
1     Self-employed          Rural             202.21   NaN     never smoked   
2           Private          Rural             105.92  32.5     never smoked   
3           Private          Urban             171.23  34.4           smokes   
4     Self-employed          Rural             174.12  24.0     never smoked   
...             ...            ...                ...   ...              ...   
5105        Private          Urban              83.75   NaN     never smoked   
5106  Self-employed          Urban             125.20  40.0     never smoked   
5107  Self-employed          Rural              82.99  30.6     never smoked   
5108        Private          Rural             166.29  25.6  formerly smoked   
5109       Govt_job          Urban              85.28  26.2          Unknown   

      stroke  
0          1  
1          1  
2          1  
3          1  
4          1  
...      ...  
5105       0  
5106       0  
5107       0  
5108       0  
5109       0  

[5110 rows x 12 columns]

**Data Cleaning**

In [ ]:
# Caluclating the sum of missing values 
df.isnull().sum()

id                     0
gender                 0
age                    0
hypertension           0
heart_disease          0
ever_married           0
work_type              0
Residence_type         0
avg_glucose_level      0
bmi                  201
smoking_status         0
stroke                 0
dtype: int64

**We have learnt that the 'bmi' attribute has 201 missing values.**<br>
We need to resolve this issue before we go any further.

In [ ]:
# Dropping the null values in 'BMI' coulmn
df.dropna(subset=['bmi',],inplace = True)
df

id  gender   age  hypertension  heart_disease ever_married  \
0      9046    Male  67.0             0              1          Yes   
2     31112    Male  80.0             0              1          Yes   
3     60182  Female  49.0             0              0          Yes   
4      1665  Female  79.0             1              0          Yes   
5     56669    Male  81.0             0              0          Yes   
...     ...     ...   ...           ...            ...          ...   
5104  14180  Female  13.0             0              0           No   
5106  44873  Female  81.0             0              0          Yes   
5107  19723  Female  35.0             0              0          Yes   
5108  37544    Male  51.0             0              0          Yes   
5109  44679  Female  44.0             0              0          Yes   

          work_type Residence_type  avg_glucose_level   bmi   smoking_status  \
0           Private          Urban             228.69  36.6  formerly smoked   
2           Private          Rural             105.92  32.5     never smoked   
3           Private          Urban             171.23  34.4           smokes   
4     Self-employed          Rural             174.12  24.0     never smoked   
5           Private          Urban             186.21  29.0  formerly smoked   
...             ...            ...                ...   ...              ...   
5104       children          Rural             103.08  18.6          Unknown   
5106  Self-employed          Urban             125.20  40.0     never smoked   
5107  Self-employed          Rural              82.99  30.6     never smoked   
5108        Private          Rural             166.29  25.6  formerly smoked   
5109       Govt_job          Urban              85.28  26.2          Unknown   

      stroke  
0          1  
2          1  
3          1  
4          1  
5          1  
...      ...  
5104       0  
5106       0  
5107       0  
5108       0  
5109       0  

[4909 rows x 12 columns]

**Outliers detection and correction**

In [ ]:
# function to detect outliers
def Outliers_Dtetector(attribute, attribute_name):
   Q1 = np.percentile(attribute, 25, interpolation = 'midpoint') 
   Q3 = np.percentile(attribute, 75, interpolation = 'midpoint') 
   IQR = Q3 - Q1
   lower_fourth = Q1 - (1.5 * IQR)
   upper_fourth = Q3 + (1.5 * IQR)
   outliers = attribute[((attribute < lower_fourth) | (attribute > upper_fourth))]
   
   print(f'Name:{attribute_name}')
   print(f'Number of Outliers: {len(outliers)}')
   print(f'Statistical Lower fourth: {lower_fourth:.3f}')
   print(f'Statistical Upper fourth: {upper_fourth:.3f}')

Outliers_Dtetector(df['bmi'], 'bmi')
print("")
Outliers_Dtetector(df['avg_glucose_level'], 'avg_glucose_level')

Name:bmi
Number of Outliers: 110
Statistical Lower fourth: 9.100
Statistical Upper fourth: 47.500

Name:avg_glucose_level
Number of Outliers: 567
Statistical Lower fourth: 22.320
Statistical Upper fourth: 168.320


In [ ]:
# variables to store median values of bmi and 
median_bmi = np.nanmedian(df['bmi'])
median_avg_glucose_level = np.nanmedian(df['avg_glucose_level'])

In [ ]:
# Replacing outliers with median values
df['bmi'].iloc[(df['bmi'] > 46.29) | (df['bmi'] < 10.30)] = median_bmi
df['avg_glucose_level'].iloc[(df['avg_glucose_level'] > 169.35) | (df['avg_glucose_level'] < 21.98)] = median_avg_glucose_level

/usr/local/lib/python3.8/dist-packages/pandas/core/indexing.py:1732: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_single_block(indexer, value, name)


In [ ]:
#dropping the column id
df = df.drop('id', axis=1)
df

gender   age  hypertension  heart_disease ever_married      work_type  \
0       Male  67.0             0              1          Yes        Private   
2       Male  80.0             0              1          Yes        Private   
3     Female  49.0             0              0          Yes        Private   
4     Female  79.0             1              0          Yes  Self-employed   
5       Male  81.0             0              0          Yes        Private   
...      ...   ...           ...            ...          ...            ...   
5104  Female  13.0             0              0           No       children   
5106  Female  81.0             0              0          Yes  Self-employed   
5107  Female  35.0             0              0          Yes  Self-employed   
5108    Male  51.0             0              0          Yes        Private   
5109  Female  44.0             0              0          Yes       Govt_job   

     Residence_type  avg_glucose_level   bmi   smoking_status  stroke  
0             Urban              91.68  36.6  formerly smoked       1  
2             Rural             105.92  32.5     never smoked       1  
3             Urban              91.68  34.4           smokes       1  
4             Rural              91.68  24.0     never smoked       1  
5             Urban              91.68  29.0  formerly smoked       1  
...             ...                ...   ...              ...     ...  
5104          Rural             103.08  18.6          Unknown       0  
5106          Urban             125.20  40.0     never smoked       0  
5107          Rural              82.99  30.6     never smoked       0  
5108          Rural             166.29  25.6  formerly smoked       0  
5109          Urban              85.28  26.2          Unknown       0  

[4909 rows x 11 columns]

**Label Encoding**

In [ ]:
# Import preprocessing from sklearn for Label Encoding
from sklearn import preprocessing
label_encoder = preprocessing.LabelEncoder()


# Encode labels in column 'gender'.
# Male : 1 , Female : 0
df['gender']= label_encoder.fit_transform(df['gender'])


# Encode labels in column 'ever_married'.
# Yes : 1 , No : 0
df['ever_married']= label_encoder.fit_transform(df['ever_married'])


# Encode labels in column 'Residence_type'.
# Urban : 1, Rural : 0
df['Residence_type']= label_encoder.fit_transform(df['Residence_type'])


# Encode labels in column 'work_type'.
# Govt_job : 0, Never_worked : 1,Private : 2,Self-employed : 3  children : 4, 
df['work_type']= label_encoder.fit_transform(df['work_type'])


# Encode labels in column 'smoking_status'.
# Unknown : 0, formerly smoked : 1, never smoked : 2, smokes : 3
df['smoking_status']= label_encoder.fit_transform(df['smoking_status'])


df

gender   age  hypertension  heart_disease  ever_married  work_type  \
0          1  67.0             0              1             1          2   
2          1  80.0             0              1             1          2   
3          0  49.0             0              0             1          2   
4          0  79.0             1              0             1          3   
5          1  81.0             0              0             1          2   
...      ...   ...           ...            ...           ...        ...   
5104       0  13.0             0              0             0          4   
5106       0  81.0             0              0             1          3   
5107       0  35.0             0              0             1          3   
5108       1  51.0             0              0             1          2   
5109       0  44.0             0              0             1          0   

      Residence_type  avg_glucose_level   bmi  smoking_status  stroke  
0                  1              91.68  36.6               1       1  
2                  0             105.92  32.5               2       1  
3                  1              91.68  34.4               3       1  
4                  0              91.68  24.0               2       1  
5                  1              91.68  29.0               1       1  
...              ...                ...   ...             ...     ...  
5104               0             103.08  18.6               0       0  
5106               1             125.20  40.0               2       0  
5107               0              82.99  30.6               2       0  
5108               0             166.29  25.6               1       0  
5109               1              85.28  26.2               0       0  

[4909 rows x 11 columns]